org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/adapters/WorkDoneProgressNotificationAdapter.java

```diff
@@ -0,0 +1,175 @@
+			return null;
+		}
+		
+		in.beginObject();
+		String kind = null, message = null, title = null;
+		Boolean cancellable = null;
+		Integer percentage = null;
+		try {
+			
+			while (in.hasNext()) {
+				String name = in.nextName();
+				switch (name) {
+				case "kind": {
+					kind = in.nextString();
+					break;
+				}
+				case "title": {
+					title = in.nextString();
+					break;
+				}
+				case "message": {
+					message = in.nextString();
+					break;
+				}
+				case "cancellable": {
+					cancellable = in.nextBoolean();
+					break;
+				}
+				case "percentage": {
+					percentage = in.nextInt();
+					break;
+				}
+				default:
+					in.skipValue();
+				}
+			}
+			in.endObject();
+			return createNotification(kind, message, title, cancellable, percentage);
+			
+		} catch (JsonSyntaxException | MalformedJsonException | EOFException exception) {
+			if (kind != null) {
+				throw new JsonParseException(exception);
+			} else {
+				throw exception;
+			}
+		}
+	}
+
+	private WorkDoneProgressNotification createNotification(String kind, String message, String title,
+			Boolean cancellable, Integer percentage) throws MalformedJsonException {
+		if(kind == null) {
+			throw new MalformedJsonException("Kind of progress notification is empty");
+		}
+		switch (kind) {
+		case "begin":
+			WorkDoneProgressBegin begin = new WorkDoneProgressBegin();
+			begin.setMessage(message);
+			begin.setCancellable(cancellable);
+			begin.setPercentage(percentage);
+			begin.setTitle(title);
+			return begin;
+		case "report":
+			WorkDoneProgressReport report = new WorkDoneProgressReport();
+			report.setMessage(message);
+			report.setCancellable(cancellable);
+			report.setPercentage(percentage);
+			return report;
+		case "end":
+			WorkDoneProgressEnd end = new WorkDoneProgressEnd();
+			end.setMessage(message);
+			return end;
+		default:
+			throw new MalformedJsonException("Kind of progress notification is unknown :"+kind);
+		}
+	}
+
+	@Override
+	public void write(JsonWriter out, WorkDoneProgressNotification notif) throws IOException {
+		out.beginObject();
+		out.name("kind");
+		out.value(notif.getKind().toString());
+		
+		switch (notif.getKind()) {
+		case begin:
+			WorkDoneProgressBegin begin = (WorkDoneProgressBegin) notif;
+			out.name("title");
+			out.value(begin.getTitle());
+			out.name("cancellable");
+			out.value(begin.getCancellable());
+			out.name("message");
+			out.value(begin.getMessage());
+			out.name("percentage");
+			out.value(begin.getPercentage());
+			break;
+		case report:
+			WorkDoneProgressReport report = (WorkDoneProgressReport) notif;
+			out.name("cancellable");
+			out.value(report.getCancellable());
+			out.name("message");
+			out.value(report.getMessage());
+			out.name("percentage");
+			out.value(report.getPercentage());
+			break;
+		case end:
+			WorkDoneProgressEnd end = (WorkDoneProgressEnd) notif;
+			out.name("message");
+			out.value(end.getMessage());
+			break;
+
+		default:
+			break;
```

Comment:
Throw exception in case there will be new notification kinds I'd like to throw it into our face if the adapter is not updated accordingly.